src/contracts/Protocol/Vault.sol

```diff
@@ -137,37 +141,45 @@ contract Vault is Ownable, OnlyCostController {
     function withdrawMetric(uint256 questionId, STAGE stage) external {
         // Checks if Metric is withdrawn for a valid stage.
         if (uint8(stage) >= 4) revert InvalidStage();
-        // Checks that only the depositer can withdraw the metric
-        if (_msgSender() != lockedMetric[questionId][stage][_msgSender()].user) revert NotTheDepositor();
-        // Checks that the metric to withdraw is not 0
-        if (lockedMetric[questionId][stage][_msgSender()].status != STATUS.DEPOSITED) revert NoMetricDeposited();
 
         if (stage == STAGE.CREATE_AND_VOTE) {
             // Checks that the question is published
             if (questionStateController.getState(questionId) != uint256(IQuestionStateController.STATE.PUBLISHED)) revert QuestionNotPublished();
 
             // Accounting & changes
-            uint256 toWithdraw = lockedMetric[questionId][stage][_msgSender()].amount;
-
-            lockedMetric[questionId][stage][_msgSender()].status = STATUS.WITHDRAWN;
-            lockedMetric[questionId][stage][_msgSender()].amount = 0;
-
-            lockedMetricByQuestion[questionId] -= toWithdraw;
-            totalLockedInVaults[_msgSender()] -= toWithdraw;
-
-            // Transfers Metric from the vault to the user.
-            metric.transfer(_msgSender(), toWithdraw);
-
-            emit Withdraw(_msgSender(), toWithdraw);
+            vaultAccounting(questionId, STAGE.CREATE_AND_VOTE);
         } else if (stage == STAGE.UNVOTE) {
-            // reimburse unvoting
+            // Check that user has a voting index, has not voted and the question state is VOTING.
+            if (questionStateController.getHasUserVoted(_msgSender(), questionId) == true) revert UserHasNotUnvoted();
+            if (questionStateController.getState(questionId) != uint256(IQuestionStateController.STATE.VOTING)) revert QuestionNotInVoting();
+
+            // Accounting & changes
+            vaultAccounting(questionId, STAGE.CREATE_AND_VOTE);
         } else if (stage == STAGE.CLAIM_AND_ANSWER) {
             // if (submissionPeriod == active) revert SubmissionPeriodActive();
         } else {
             // if (reviewPeriod == active) revert ReviewPeriodActive();
         }
     }
 
+    function vaultAccounting(uint256 questionId, STAGE stage) internal {
+        if (_msgSender() != lockedMetric[questionId][stage][_msgSender()].user) revert NotTheDepositor();
+        if (lockedMetric[questionId][stage][_msgSender()].status != STATUS.DEPOSITED) revert NoMetricDeposited();
+
+        uint256 toWithdraw = lockedMetric[questionId][stage][_msgSender()].amount;
```

Comment:
could this get funky if the user has lockedMetric for different reasons?  like when they create a question it gets an auto-upvote -if they unvote, would it reimburse their creation fees?

Reply:
Yep you're right, good catch. Fixing.